de.metas.banking/de.metas.banking.base/src/main/java/de/metas/banking/payment/impl/BankStatmentPaymentBL.java

```diff
@@ -102,23 +113,150 @@ public void setC_Payment(IBankStatementLineOrRef lineOrRef, I_C_Payment payment)
 		// Bank Statement Line specific:
 		if (lineOrRef instanceof org.compiere.model.I_C_BankStatementLine)
 		{
-			org.compiere.model.I_C_BankStatementLine bsl = (org.compiere.model.I_C_BankStatementLine)lineOrRef;
+			final org.compiere.model.I_C_BankStatementLine bsl = (org.compiere.model.I_C_BankStatementLine)lineOrRef;
 			bsl.setDescription(payment.getDescription());
 		}
 	}
 
-	public void setPayAmt(IBankStatementLineOrRef lineOrRef, BigDecimal payAmt)
+	@Override
+	public void findOrCreateUnreconciledPaymentsAndLinkToBankStatementLine(final de.metas.banking.model.I_C_BankStatementLine line)
+	{
+		final boolean manualActionRequired = findAndLinkPaymentToBankStatementLineIfPossible(line);
+
+		if (!manualActionRequired)
+		{
+			setOrCreateAndLinkPaymentToBankStatementLine(line, null);
+		}
+	}
+
+	/**
+	 * @return true if the automatic flow should STOP as manual action is required; false if the automatic flow should continue
+	 */
+	private boolean findAndLinkPaymentToBankStatementLineIfPossible(final de.metas.banking.model.I_C_BankStatementLine line)
+	{
+		// a payment is already linked
+		if (line.getC_Payment_ID() > 0)
+		{
+			return true;
+		}
+		if (line.getC_BPartner_ID() <= 0)
+		{
+			return true;
+		}
+
+		final boolean isReceipt = line.getStmtAmt().signum() >= 0;
+		final BigDecimal expectedPaymentAmount = isReceipt ? line.getStmtAmt() : line.getStmtAmt().negate();
+
+		final Money money = Money.of(expectedPaymentAmount, CurrencyId.ofRepoId(line.getC_Currency_ID()));
+		final BPartnerId bPartnerId = BPartnerId.ofRepoId(line.getC_BPartner_ID());
+		final ImmutableSet<PaymentId> possiblePayments = Services.get(IPaymentDAO.class).retrieveAllMatchingPayments(isReceipt, bPartnerId, money);
+
+		// Don't create a new Payment and don't link any of the existing payments if there are multiple payments found.
+		// The user must fix this case manually by choosing the correct Payment
+		if (possiblePayments.size() > 1)
+		{
+			return true;
+		}
+
+		if (possiblePayments.size() == 1)
+		{
+			line.setC_Payment_ID(possiblePayments.iterator().next().getRepoId());
+			InterfaceWrapperHelper.save(line);
+		}
+		return false;
+	}
+
+	@Override
+	public Optional<PaymentId> setOrCreateAndLinkPaymentToBankStatementLine(@NonNull final de.metas.banking.model.I_C_BankStatementLine line, @Nullable final PaymentId paymentIdToSet)
+	{
+		// a payment is already linked
+		if (line.getC_Payment_ID() > 0)
+		{
+			return Optional.of(PaymentId.ofRepoId(line.getC_Payment_ID()));
+		}
+
+		if (paymentIdToSet != null)
+		{
+			final I_C_Payment payment = Services.get(IPaymentDAO.class).getById(paymentIdToSet);
+			setC_Payment(line, payment);
+
+			InterfaceWrapperHelper.save(line);
+			InterfaceWrapperHelper.save(payment);
```

Comment:
consider saving the `payment ` in some payment DAO/repo class

Reply:
done